### PR TITLE
[Golang][client] fix for schema definition name `file`

### PR DESCRIPTION
**Pull request by grokify**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [x] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@antihax
@bvwells

There are some open items listed at the end of this post so it is not yet ready for merging. Please review the general approach.

### Description of the PR

When generating a client with a v3.0 spec Schema or v2.0 spec Definition named `File`, the generator will convert the type to `*os.File` instead of the model named `File`. This is a generic issue where a model schema/definition name may collide with a `typeMapping` key.

An example spec excerpt looks like:

```yaml
    PoolResponseResource:
      type: "object"
      properties:
          # ...
        sourceFiles:
          type: "array"
          items:
            $ref: "#/definitions/File"
    File:
      type: "object"
      properties:
        sourceFile:
          type: "string"
        originalFileName:
          type: "string"
```

In the above, `PoolResponseResource` will have a property `SourceFiles` with type `[]*os.File` instead of the desired `[]File`.

This occurs because abstract class `AbstractGoCodegen` converts `File` to `*os.File` using `typeMapping`:

```java
        typeMapping.put("File", "*os.File");
        typeMapping.put("file", "*os.File");
```

`getSchemaType` and `getTypeDeclaration` both check against `typeMapping`, e.g.:

```java
if (typeMapping.containsKey(openAPIType)) {
            type = typeMapping.get(openAPIType);
```

`getSchemaType` gets a baseline type to process from `super.getSchemaType` which returns a single string that appears to represent 3 things: (a) definition/schema names, (b) types, and (c) formats. `typeMapping` covers the following:

| schema type | language type | spec (2.0) |
|----------------|------------------|-------|
| `integer` | `int32` | `type` only: `"type": "integer"` |
| `long` | `int64` | `type` and `format`: `"type": "integer", "format": "int64"` |
| `File` | `*os.File` | `$ref` only: `"$ref": "#/definitions/File"` |

#### Proposed Fix

This PR includes a minimal fix and test case to consider the value of `$ref` when determining the type. If `$ref` is present and matches the Spec 2.0/3.0 definition/schema path, `typeMapping `is not used.

#### Notes

The added test, `filePropertyTest` in `GoModelTest.java` uses the spec 2.0 model reference: `#/definitions/File`, however, when building a client, it appears a 2.0 spec using `#/definitions/File` is converted to a 3.0 spec `#/components/schemas/File` by `super.getSchemaType`

#### References

* https://github.com/grokify/go-ringcentral/issues/26

#### ~~Open Items / Next Steps~~

1. ~~This update generates client code conditionally if a spec definition with name `File` (or other schema name with typeMapping collision) is present, so an additional fake PetStore test endpoint will be provided to test the results.~~ `filePropertyTest` added to `GoModelTest.java`.
1. ~~This currently generates a regression issue that needs to be resolved. A struct will be created with name `ProfileImageInfoUri` but be referenced as `ProfileImageInfoURI`. This is not covered in the Petstore tests, but appears using the RingCentral spec. Will resolve and look to add a test case for this.~~ Fixed in e51623bb9bca3592a27978567921e56e4dd6a7c9 . Covered in `filePropertyTest` added to `GoModelTest.java`.

### CODE REVIEW

**Comment by grokify** [comment]
Build errors so far:

* Shippable: Failing due to `ShutdownMonitorThread already started` race condition
* CircleCI: Worked previously. Recently failed due to JAR download error

**Comment by antihax** [comment]
Thanks for looking into this.

Is this a common problem that could affect other languages?

**Comment by wing328** [comment]
I've restarted the shippable job and it's fine now.

For CircleCI, seems like there're some changes and I could no longer restart it: https://circleci.com/gh/grokify/openapi-generator/28?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

The failure has nothing to do with the change and the Go test passed:
```
[ERROR] Error fetching link: /home/circleci/OpenAPITools/openapi-generator/modules/openapi-generator/home/circleci/OpenAPITools/openapi-generator/modules/openapi-generator-cli/target/apidocs/package-list. Ignored it.
line 2 column 2 - Error: <additionalproperties> is not recognized!
line 3 column 6 - Error: <additionalproperty> is not recognized!
line 2 column 2 - Error: <additionalproperties> is not recognized!
line 3 column 6 - Error: <additionalproperty> is not recognized!
=== RUN   TestOAuth2
--- PASS: TestOAuth2 (0.29s)
=== RUN   TestBasicAuth
--- PASS: TestBasicAuth (0.00s)
=== RUN   TestAccessToken
--- PASS: TestAccessToken (0.00s)
=== RUN   TestAPIKeyNoPrefix
--- PASS: TestAPIKeyNoPrefix (0.00s)
=== RUN   TestAPIKeyWithPrefix
--- PASS: TestAPIKeyWithPrefix (0.00s)
=== RUN   TestDefaultHeader
--- PASS: TestDefaultHeader (0.00s)
=== RUN   TestHostOverride
--- PASS: TestHostOverride (0.01s)
=== RUN   TestAddPet
--- PASS: TestAddPet (0.00s)
=== RUN   TestFindPetsByStatusWithMissingParam
--- PASS: TestFindPetsByStatusWithMissingParam (0.00s)
=== RUN   TestGetPetById
--- PASS: TestGetPetById (0.00s)
=== RUN   TestGetPetByIdWithInvalidID
--- PASS: TestGetPetByIdWithInvalidID (0.00s)
=== RUN   TestUpdatePetWithForm
--- PASS: TestUpdatePetWithForm (0.00s)
=== RUN   TestFindPetsByTag
--- PASS: TestFindPetsByTag (0.00s)
=== RUN   TestFindPetsByStatus
--- PASS: TestFindPetsByStatus (0.00s)
=== RUN   TestUploadFile
--- PASS: TestUploadFile (0.03s)
=== RUN   TestUploadFileRequired
--- PASS: TestUploadFileRequired (0.00s)
=== RUN   TestDeletePet
--- PASS: TestDeletePet (0.00s)
=== RUN   TestPlaceOrder
--- PASS: TestPlaceOrder (0.02s)
=== RUN   TestCreateUser
--- PASS: TestCreateUser (0.01s)
=== RUN   TestCreateUsersWithArrayInput
--- PASS: TestCreateUsersWithArrayInput (0.01s)
=== RUN   TestGetUserByName
--- PASS: TestGetUserByName (0.00s)
=== RUN   TestGetUserByNameWithInvalidID
--- PASS: TestGetUserByNameWithInvalidID (0.00s)
=== RUN   TestUpdateUser
--- PASS: TestUpdateUser (0.00s)
=== RUN   TestDeleteUser
--- PASS: TestDeleteUser (0.00s)
PASS
ok  	_/home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/go	0.405s
```

**Comment by grokify** [comment]
> Is this a common problem that could affect other languages?

I think so if the general approach is the same. Every generator that called `getSchemaType` and performed language type conversion may run into this issue since they would need to separately implement a check for schema/definition name vs. built-in data type.

It seems like a more generic and less error-prone solution would be to have a method like `getSchemaType` across all generators that returned additional information on whether the returned string was a built-in data type or a schema/definition name so the generators could process them differently without implementing individual checks. I thought about implementing this approach as will but since this is a deeper change, it's worth a discussion.

**Comment by grokify** [comment]
All tests are passing now.

File: modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml [hunk 1]
```diff
@@ -67,7 +67,6 @@ paths:
           description: Status values that need to be considered for filter
           required: true
           style: form
-          explode: false
```
Comment (wing328) [inline modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml]: FYI. Opened https://github.com/Mermade/oas-kit/issues/76 to track it.

**Comment by wing328** [comment]
The shippable error does not seem to be related to the change.

**Pull request merged**